### ISSUE TITLE
two ranks for each triple

### ISSUE THREAD

**Issue by Chen-Cai-OSU** [Accenture/AmpliGraph#101]
Hello,

I found that when I run the doctest in the protocol.py (ampligraph/evaulation/protocol.py), each triple has two ranks when I call evaluate_performance, but if I call model.predict, I got only one rank.

Also, it seems the rank2 is roughly the sum of rank1. Is the two number is rank1 correspond to (?, p, o) and (s, p, ?). If that's the case, why rank2 is not the exact the sum of rank1?

Here is the code:

    import sys
    import numpy as np
    from ampligraph.datasets import load_wn18
    from ampligraph.latent_features import ComplEx
    from ampligraph.evaluation import evaluate_performance

    X = load_wn18()
    model = ComplEx(batches_count=10, seed=0, epochs=1, k=150, eta=10,
                 loss = 'pairwise', optimizer = 'adagrad')
    model.fit(np.concatenate((X['train'], X['valid'])))
    filter = np.concatenate((X['train'], X['valid'], X['test']))
    rank1 = evaluate_performance(X['test'][:1], model=model, filter_triples=filter)
    print(rank1)
    rank2 = model.predict(X['test'][:1], get_ranks=True)
    print(rank2)

**Comment by Chen-Cai-OSU** [comment]
I am a bit confused by that's the case since evaluate_performance is basically calling model. predict function.

**Comment by sumitpai** [comment]
**evaluate_performance API:**

The ranks returned by `evaluate_performance` depends on several things: the protocol being used, sides being corrupted and whether the filtering performed.

In the case above, `use_default_protocol` is `True`(default), and you are using a filter. (When using default protocol, `corrupt_side` is always set to `s+o`) - more details below.

In `evaluate_performance` when `corrupt_side` is set to `s+o` there are two cases:

- One is when `use_default_protocol` is set to `True` , where the < ?, p, o > and  < s, p, ? > corruptions are generated and ranked separately (i.e. we don't append them before ranking). For example, if there are 100 distinct entities in the KG, then we would generate 100 different corruptions for < ?, p, o > and 100 for < s, p, ? >. While ranking, these corruptions are ranked separately (so rank will be between 1<=rank<=100 for the test triple - one for < ?, p, o >  and other for < s, p, o > ). So in your case you are seeing 2 different ranks as output of evaluate_performance.

- In the second case where `use_default_protocol` is set to `False`, we concatenate < ?, p, o > and  < s, p, ? > corruptions and then they are ranked 1<=rank<=200. In this case you would get a single rank. 

In the research papers, the reported MRRs are on the ranks computed using default protocol - that is why we have the use_default_protocol mode.

**predict API:**

When you call model.predict it does the second case (in an unfiltered setting). That's why you get a single rank.  

The `predict` API is meant to be used only to score triples. If you really want to perform rigorous evaluation, we would suggest you to use `evaluate_performance`

To get the same ranks as evaluate_performance with predict API (i.e. using default protocol in filtered setting), you can do the following:

```
from ampligraph.evaluation import to_idx

#map to internal ids
filter_mappings = to_idx(filter, ent_to_idx=model.ent_to_idx, rel_to_idx=model.rel_to_idx)
#set the filter
model.set_filter_for_eval(filter_mappings)

#Create a configuration for evaluation
eval_dict = {}
eval_dict['default_protocol'] = True
eval_dict['corrupt_side'] = 's+o'
#configure the settings
model.configure_evaluation_protocol(eval_dict)

rank2 = model.predict(X['test'][:1], get_ranks=True)
print(rank2)
```

In summary, use `predict` API if you want to score triples for some of your use cases. But when you want to rigorously evaluate triples use `evaluate_performance` API. `evaluate_performance` can be thought of as an **high level API** which helps a user to perform evaluation with several options.
